core/trino-main/src/main/java/io/trino/execution/scheduler/StageEventDrivenTaskSourceFactory.java

```diff
@@ -0,0 +1,1310 @@
+        private final Map<PlanFragmentId, PlanNodeId> remoteSources;
+        private final Supplier<Map<PlanNodeId, SplitSource>> splitSourceSupplier;
+        private final SplitAssigner assigner;
+        private final EventDrivenTaskSource.Callback callback;
+        private final Executor executor;
+        private final int splitBatchSize;
+        private final long targetPartitionSizeInBytes;
+        private final FaultTolerantPartitioningScheme sourcePartitioningScheme;
+        private final LongConsumer getSplitTimeRecorder;
+        private final SetMultimap<PlanNodeId, PlanFragmentId> remoteSourceFragments;
+
+        @GuardedBy("this")
+        private boolean started;
+        @GuardedBy("this")
+        private boolean closed;
+        @GuardedBy("this")
+        private final Closer closer = Closer.create();
+
+        private final Object assignerLock = new Object();
+
+        @VisibleForTesting
+        StageTaskSource(
+                Map<PlanFragmentId, Exchange> sourceExchanges,
+                Map<PlanFragmentId, PlanNodeId> remoteSources,
+                Supplier<Map<PlanNodeId, SplitSource>> splitSourceSupplier,
+                SplitAssigner assigner,
+                Callback callback,
+                Executor executor,
+                int splitBatchSize,
+                long targetPartitionSizeInBytes,
+                FaultTolerantPartitioningScheme sourcePartitioningScheme,
+                LongConsumer getSplitTimeRecorder)
+        {
+            this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
+            this.remoteSources = ImmutableMap.copyOf(requireNonNull(remoteSources, "remoteSources is null"));
+            checkArgument(
+                    sourceExchanges.keySet().equals(remoteSources.keySet()),
+                    "sourceExchanges and remoteSources are expected to contain the same set of keys: %s != %s",
+                    sourceExchanges.keySet(),
+                    remoteSources.keySet());
+            this.splitSourceSupplier = requireNonNull(splitSourceSupplier, "splitSourceSupplier is null");
+            this.assigner = requireNonNull(assigner, "assigner is null");
+            this.callback = requireNonNull(callback, "callback is null");
+            this.executor = requireNonNull(executor, "executor is null");
+            this.splitBatchSize = splitBatchSize;
+            this.targetPartitionSizeInBytes = targetPartitionSizeInBytes;
+            this.sourcePartitioningScheme = requireNonNull(sourcePartitioningScheme, "sourcePartitioningScheme is null");
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            remoteSourceFragments = remoteSources.entrySet().stream()
+                    .collect(toImmutableSetMultimap(Map.Entry::getValue, Map.Entry::getKey));
+        }
+
+        @Override
+        public synchronized void start()
```

Comment:
`startIfNotStarted`?

Reply:
Also maybe split the method into two - one for creating `SplitLoader` and other for starting those up with top level exception handlingn.
The indentation depth is below my comfort level right now.

Reply:
> startIfNotStarted?

We usually call it just `start` in other places. Do you think this one should be different?

Reply:
> Also maybe split the method into two

I was trying to do this, but since

```
Set<PlanFragmentId> finishedFragments = new HashSet<>();
            Set<PlanNodeId> allSources = new HashSet<>();
            Set<PlanNodeId> finishedSources = new HashSet<>();
```

have to be shared it felt like splitting it makes it less readable. Happy to discuss offline.

Reply:
I think (maybe this is not really the case) that typically a `start()` method would throw if "started object" is already started. Not just ignore the call.

Reply:
Hmm, we can do that. I don't know why did I implement it to simply ignore it. I don't think it is ever called more than once.